src/communication/response_broker.hpp

```diff
@@ -195,61 +230,66 @@ namespace cubcomm
       std::lock_guard<std::mutex> lockg (m_mutex);
 
       payload_or_error_type &ent = m_response_payloads[a_rsn];
+      assert (!ent.m_response_or_error_present);
       ent.m_error = std::move (a_error);
+      ent.m_response_or_error_present = true;
     }
     // notify all because there is more than one thread waiting for data on the same bucket
     // ideally, with an adequately sized bucket pool, the contention should be minimal
     m_condvar.notify_all ();
   }
 
-  template <typename T_PAYLOAD, typename T_ERROR>
-  std::tuple<T_PAYLOAD, T_ERROR>
-  response_broker<T_PAYLOAD, T_ERROR>::get_response (response_sequence_number a_rsn)
-  {
-    return get_bucket (a_rsn).get_response (a_rsn);
-  }
-
   template <typename T_PAYLOAD, typename T_ERROR>
   std::tuple<T_PAYLOAD, T_ERROR>
   response_broker<T_PAYLOAD, T_ERROR>::bucket::get_response (response_sequence_number a_rsn)
   {
     constexpr std::chrono::milliseconds millis_100 { 100 };
 
     {
-      typename response_payload_container_type::iterator found_it { m_response_payloads.end () };
+      std::unique_lock<std::mutex> ulock (m_mutex);
+
+      // nobody else but us should access this payload container but must be accessed locked
+      typename response_payload_container_type::iterator found_it = m_response_payloads.find (a_rsn);
+      payload_or_error_type &ent = (*found_it).second;
 
-      auto condvar_pred = [this, &found_it, a_rsn] ()
+      auto condvar_pred = [&ent] ()
       {
-	found_it = m_response_payloads.find (a_rsn);
-	return (found_it != m_response_payloads.end ());
+	return ent.m_response_or_error_present;
       };
 
-      std::unique_lock<std::mutex> ulock (m_mutex);
       // a way out in case neither value nor error is registered as a response
       // which can happen in case - eg - that the connection is dropped
       while (!m_terminate)
 	{
 	  if (m_condvar.wait_for (ulock, millis_100, condvar_pred))
 	    {
-	      assert (found_it != m_response_payloads.end ());
+	      assert (ent.m_response_or_error_present);
 	      std::tuple<T_PAYLOAD, T_ERROR> payload_or_error
 	      {
-		std::move ((*found_it).second.m_payload ),
-		std::move ((*found_it).second.m_error)
+		std::move (ent.m_payload ),
+		std::move (ent.m_error)
 	      };
+
 	      m_response_payloads.erase (found_it);
 	      return payload_or_error;
 	    }
+
+	  // NOTE: when terminate is invoked there is the possibility to implement a configurable behaviour:
+	  //  - either abort and return error - as done below
+	  //  - or continue waiting - with a timeout - until the request is serviced by the peer server
 	}
     }
 
+    // if here, terminate was specified and there is at most one thread to notify
```

Comment:
Yes. Thank you for this!
I have also adapted and extended the unit test for the addition of the `register_request` function.
Indeed, without emptying the bucket container, it would not work.
This is easy to see with the unit test - that's why I added an explicit unit test for this - "Test response broker without responses and with terminate/abort".

While developing, I tried to replicate this behavior by instrumenting the PS code with a call to `request_sync_client_server::stop_incoming_communication_thread` and logging - but I was not able to.